url6a9f553762_Bresiu_socket_io_tgz-pJ8-MainJ8/src/com/github/nkzawa/socketio/client/Socket.java

```diff
@@ -0,0 +1,429 @@
+        packet.nsp = this.nsp;
+        this.io.packet(packet);
+    }
+
+    private void onopen() {
+        logger.fine("transport is open - connecting");
+
+        if (!"/".equals(this.nsp)) {
+            this.packet(new Packet(Parser.CONNECT));
+        }
+    }
+
+    private void onclose(String reason) {
+        logger.fine(String.format("close (%s)", reason));
+        this.connected = false;
+        this.disconnected = true;
+        this.emit(EVENT_DISCONNECT, reason);
+    }
+
+    private void onpacket(Packet packet) {
+        if (!this.nsp.equals(packet.nsp)) return;
+
+        switch (packet.type) {
+            case Parser.CONNECT:
+                this.onconnect();
+                break;
+
+            case Parser.EVENT:
+                this.onevent(packet);
+                break;
+
+            case Parser.BINARY_EVENT:
+                this.onevent(packet);
+                break;
+
+            case Parser.ACK:
+                this.onack(packet);
+                break;
+
+            case Parser.BINARY_ACK:
+                this.onack(packet);
+                break;
+
+            case Parser.DISCONNECT:
+                this.ondisconnect();
+                break;
+
+            case Parser.ERROR:
+                this.emit(EVENT_ERROR, packet.data);
+                break;
+        }
+    }
+
+    private void onevent(Packet<JSONArray> packet) {
+        List<Object> args = new ArrayList<Object>(Arrays.asList(toArray(packet.data)));
+        logger.fine(String.format("emitting event %s", args));
+
+        if (packet.id >= 0) {
+            logger.fine("attaching ack callback to event");
+            args.add(this.ack(packet.id));
+        }
+
+        if (this.connected) {
+            String event = (String)args.remove(0);
+            super.emit(event, args.toArray());
+        } else {
+            this.receiveBuffer.add(args);
+        }
+    }
+
+    private Ack ack(final int id) {
+        final Socket self = this;
+        final boolean[] sent = new boolean[] {false};
+        return new Ack() {
+            @Override
+            public void call(final Object... args) {
+                EventThread.exec(new Runnable() {
+                    @Override
+                    public void run() {
+                        if (sent[0]) return;
+                        sent[0] = true;
+                        logger.fine(String.format("sending ack %s", args.length != 0 ? args : null));
+
+                        int type = HasBinaryData.hasBinary(args) ? Parser.BINARY_ACK : Parser.ACK;
+                        Packet<JSONArray> packet = new Packet<JSONArray>(type, new JSONArray(Arrays.asList(args)));
+                        packet.id = id;
+                        self.packet(packet);
+                    }
+                });
+            }
+        };
+    }
+
+    private void onack(Packet<JSONArray> packet) {
+        logger.fine(String.format("calling ack %s with %s", packet.id, packet.data));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

This code uses &#39;%s&#39; to format int: packet.id. This is a potential locale-sensitive handling issue. It might cause errors in the handling and processing of the statement at line: 338. Consider formatting this data with &#39;%d&#39; instead.